### PR TITLE
Feature 2227 namespace

### LINKED ISSUE #2227: Remove namespace specification from header files
## Describe the Enhancement ##
This issue was raised on 8/8/2022 during a Modern C++ weekly meeting. Please see the following discussion:
https://stackoverflow.com/questions/1265039/using-std-Namespace

Based on this, `using namespace ...` should never be included in a header file. However, of the 467 header files in the MET codebase, `using namespace std` and `using namespace netCDF` appears in 39 and 32 of them, respectively.

This task is to remove them from the header files and make any updates required to ensure the code compiles and all existing tests run.

### Time Estimate ###
2 days?

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
Non needed.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
Per Tara, split across keys

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
No impacts.

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [x] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [x] Push local changes to GitHub.
- [x] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [x] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [x] Iterate until the reviewer(s) accept and merge your changes.
- [x] Delete your fork or branch.
- [x] Close this issue.

### PR DESCRIPTION
**Pull request by jprestop**
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>
If **yes**, please describe:</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
Compiled on seneca and via GitHub Actions

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
Review changes if desired.  Ensure that all tests pass.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[None needed]**

- [x] Do these changes include sufficient testing updates? **[No updates needed]**

- [x] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [x] Please complete this pull request review by **[10/10/2022]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Development** with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: src/basic/vx_config/config.tab.cc [hunk 1]
```diff
@@ -1389,7 +1389,7 @@ yyparse (void)
 
   YYDPRINTF ((stderr, "Starting parse\n"));
 
-  yystate = 8;
+  yystate = 0;
```
Comment (JohnHalleyGotway) [inline src/basic/vx_config/config.tab.cc]:
```suggestion
  yystate = 8;
```
This needs to remain a value of 8. That enables empty config files to be parsed without error. However, this is automatically generated code. So every time it gets regenerated, it resets that 8 to 0. I suppose we need to update the Makefile.am logic to switch the 0 back to 8 after regenerating it. Ugh.

Comment (jprestop) [reply src/basic/vx_config/config.tab.cc]: @JohnHalleyGotway I'm not sure why that change didn't get picked up in the pull from develop... In any case, I will make that change and commit it.  Sorry about that!

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of this PR, subject to fixing the `yystate = 8` line.

- I see that GHA ran fine and flagged not diffs.
- I checked out this feature branch and diffed files between the feature_2227_namespace and develop branches. I found the following:
- Changes to 179 '.h' header files (ignoring whitespace diffs).
```
git diff -w develop `find ./ -name "*.h"` | egrep -i '^\+'  | egrep -v '^\+\+\+' | wc -l
978
```
Of the 978 diff lines, 916 have "std::" or "netCDF::" in them, which are expected diffs.
The remaining 62 are just new blank lines or comments.
- Changes to 78 '.cc' files. Inspected the diffs and only found the issue already noted about `yystate = 8`. The changes are basically adding/removing the "std::" scoping.

**Comment by jprestop** [comment]
I updated the `yystate = 8` line and committed that change.  Thank you also for running all of those diffs and detailing that here.  I did some basic checking of the header files, but for some reason, did not think to list that under "Describe testing already performed for these changes:" even though I now realize I should have.

**Comment by JohnHalleyGotway** [comment]
@jprestop, last year you worked on this issue to remove `using namespace std` from all the MET header files (*.h). So whenever we reference an STL class in a header file, we put `std::` before it to define the scope. However, I see about 119 references to `string` that I assume should be `std::string` instead.
```
grep ' string ' `find ./ -name "*.h"` | sed -r 's%//.+%%g' | grep ' string ' | wc -l
119
``` 
For example, in `src/libcode/vx_data2d_nccf/nccf_file.h`, I assume that:
```
unixtime get_valid_time_from_file_path(const string &filepath) const;
```
Should be:
```
unixtime get_valid_time_from_file_path(const std::string &filepath) const;
```
However it still compiles without that. Any thoughts?

**Comment by hsoh-u** [comment]
"using namespace std;" statement is before "#include "nccf_file.h" at nccf_file.cc. The compile error happens once "using namespace xxx" statements are moved down below "#include" statements.

**Comment by jprestop** [comment]
Thank you @hsoh-u for pointing that out.  @JohnHalleyGotway, I see in the issue [Remove namespace specification from header files #2227](https://github.com/dtcenter/MET/issues/2227):

> Please see the following discussion:
https://stackoverflow.com/questions/1265039/using-std-Namespace

Looking at that link again now, I see:

> Herb Sutter and Andrei Alexandrescu have this to say in "Item 59: Don't write namespace usings in a header file or before an #include" of their book, C++ Coding Standards: 101 Rules, Guidelines, and Best Practices:
>>In short: You can and should use namespace using declarations and directives liberally in your implementation files after #include directives and feel good about it.

It looks like I did not add the `using namespace std` below the include statements (although perhaps for some I did based o n knowing that I did add `std::` in some places.   Should we create or reopen this issue to move `using namespace std` below the include statements?

**Pull request merged**